app/actions.py

```diff
@@ -92,30 +92,170 @@ def performDeleteRange(self, upperRow, upperCol, lowerRow, lowerCol):
+    """
+    if (bookmarkRange1[-1] >= bookmarkRange2[0] and
+        bookmarkRange1[0] <= bookmarkRange2[-1]):
+      return True
+    else:
+      return False
+
   def bookmarkAdd(self):
-    app.bookmarks.add(self.fullPath, self.penRow, self.penCol, 0,
-        app.selectable.kSelectionNone)
+    """
+    Adds a bookmark at the cursor's location. If multiple lines are 
+    selected, all existing bookmarks in those lines are overwritten 
+    with the new bookmark.
 
-  def bookmarkGoto(self):
+    Args:
+      None.
+
+    Returns:
+      None.
+    """
+    newBookmark = self.dataToBookmark()
+    self.bookmarkRemove()
+    bisect.insort(self.bookmarks, newBookmark)
+
+  def bookmarkGoto(self, bookmark):
+    """
+    Goes to the bookmark that is passed in.
+
+    Args:
+      bookmark (tuple): contains bookmarkRange and bookmarkData. More info can
+        be found in the dataToBookmark function.
+
+    Returns:
+      None.
+    """
     app.log.debug()
-    bookmark = app.bookmarks.get(self.view.bookmarkIndex)
-    if bookmark:
-      self.selectText(bookmark['row'], bookmark['col'], bookmark['length'],
-          bookmark['mode'])
+    bookmarkRange, bookmarkData = bookmark
+    cursorRow, cursorCol = bookmarkData['cursor']
+    penRow, penCol = bookmarkData['pen']
+    markerRow, markerCol = bookmarkData['marker']
+    selectionMode = bookmarkData['selectionMode']
+    self.__skipCursorScroll = True
+    self.cursorMoveAndMark(penRow - self.penRow, penCol - self.penCol,
+        markerRow - self.markerRow, markerCol - self.markerCol,
+        selectionMode - self.selectionMode)
+    self.redo()
 
   def bookmarkNext(self):
+    """
+    Goes to the closest bookmark after the cursor.
+
+    Args:
+      None.
+
+    Returns:
+      None.
+    """
     app.log.debug()
-    self.view.bookmarkIndex += 1
-    self.bookmarkGoto()
+    bookmark = None
+    _, _, lowerRow, _ = self.startAndEnd()
+    numBookmarks = len(self.bookmarks)
+    tempBookmark = ((lowerRow,float('inf')),)
+    index = bisect.bisect(self.bookmarks, tempBookmark)
+    if numBookmarks:
+      bookmark =  self.bookmarkGoto(self.bookmarks[index % numBookmarks])
+    else:
+      self.setMessage("No bookmarks to jump to")
 
   def bookmarkPrior(self):
+    """
+    Goes to the closest bookmark before the cursor.
+
+    Args:
+      None.
+
+    Returns:
+      None.
+    """
     app.log.debug()
-    self.view.bookmarkIndex -= 1
-    self.bookmarkGoto()
+    bookmark = None
+    upperRow, _, _, _ = self.startAndEnd()
+    numBookmarks = len(self.bookmarks)
+    tempBookmark = ((upperRow,),)
+    index = bisect.bisect_left(self.bookmarks, tempBookmark)
+    if numBookmarks:
+      bookmark =  self.bookmarkGoto(self.bookmarks[index - 1])
+    else:
+      self.setMessage("No bookmarks to jump to")
 
   def bookmarkRemove(self):
+    """
+    Removes bookmarks in all selected lines.
+
+    Args:
+      None.
+
+    Returns:
+      None.
```

Comment:
What do ya think about leaving out Args and Returns if there aren't any to speak of? I'm flexible on it if ya really like it this way.

Reply:
Uhh I'm still kind of experimenting and learning documentation styles. I thought about leaving it out, but thought maybe it would be more clear for skimming through functions as a reminder of what they take/return. It does take up a lot of space, but it's what I saw on Google's Python style docs so I figured I would give it a try. 

I think we could keep it to keep the format consistent, or else we should at least document it in the docstring comment to detail what the function's arguments/return values are (so someone doesn't think it's just incomplete documentation).